src/compiler/type_def.rs

```diff
@@ -39,7 +59,7 @@ pub struct TypeDef {
     ///
     /// Some expressions are infallible (e.g. the [`Literal`][crate::expression::Literal] expression, or any
     /// custom function designed to be infallible).
-    fallible: bool,
+    fallibility: Fallibility,
```

Comment:
We should document the requirement for accuracy of the `Fallibility` somewhere. Maybe on the `Expression::type_def` function, or maybe here? Something along the lines of, if the function _can_ fail, the fallibility must _not_ be `CannotFail`, and if the function _can_ succeed, the fallibility must not be `AlwaysFails`. The rest is best effort. (So always returning `MightFail` is "correct", but potentially undesirable.